devices/PAA5100JE.py

```diff
@@ -0,0 +1,255 @@
+        send_buf = bytearray([register]) + bytearray(length)
+        # Create a result buffer of the same length as the send_buf
+        result = bytearray(len(send_buf))
+        
+        if self._spi_cs_gpio:
+            self.select.on()
+        self.spi.write_readinto(send_buf, result)
+        if self._spi_cs_gpio:
+            self.select.off()
+        
+        # Return the read result, skipping the first byte (which corresponds to the register)
+        return result[1:] if length > 1 else result[1]
+
+    def _bulk_write(self, data):
+        for x in range(0, len(data), 2):
+            register, value = data[x : x + 2]
+            if register == WAIT:
+                time.sleep(value / 1000)
+            else:
+                self._write(register, value)
+                
+    def read_registers(self, reg, buf, len):
+        self.select.on()
+        self.spi.write(bytearray([reg]))
+        # Read data from the register
+        buf[:] = self.spi.read(len)
+        self.select.off()   
+        
+class MotionDetector(Analog_input):
+    def __init__(self, reset, cs1, cs2,
+                 name='MotDet', threshold=1, calib_coef=1,
+                 sampling_rate=100, event='motion'):
+        
+        # Create SPI objects
+        self.motSen_x = PAA5100JE('SPI2', reset, cs1)
+        self.motSen_y = PAA5100JE('SPI2', reset, cs2)
+
+        self._threshold = threshold
+        self.calib_coef = calib_coef
+        
+        # Motion sensor variables
+        self.x_buffer = array('i', [0, 0])
+        self.y_buffer = array('i', [0, 0])
+        self.x_buffer_mv = memoryview(self.x_buffer)
+        self.y_buffer_mv = memoryview(self.y_buffer)
+        
+        self.prev_x = 0
+        self.prev_y = 0
+        self.delta_x, self.delta_y = 0, 0    # accumulated position
+        self._delta_x, self._delta_y = 0, 0  # instantaneous position
+        self.x, self.y = 0, 0  # to be accessed from the task, unit=mm
+        
+        # Parent
+        Analog_input.__init__(self, pin=None, name=name + '-X', sampling_rate=int(sampling_rate),
+                              threshold=threshold, rising_event=event, falling_event=None,
+                              data_type='l')
+        self.data_chx = self.data_channel
+        self.data_chy = Data_channel(name + '-Y', sampling_rate, data_type='l')
+        self.crossing_direction = True  # to conform to the Analog_input syntax
+        self.timestamp = fw.current_time
+        self.acquiring = False
+        
+        gc.collect()
+        time.sleep_ms(2)
+    
+    @property
+    def threshold(self):
+        "return the value in cms"
+        return math.sqrt(int(self._threshold**2) * self.calib_coef)
+    
+    def reset_delta(self):
+        # reset the accumulated position data
+        self.delta_x, self.delta_y = 0, 0
+    
+    def read_sample(self):
+        # Units are in millimeters
+        current_motion_x = self.motSen_x.get_motion()

```

Comment:
be aware as to not allocate any memory in any code that is executed inside the interrupt routine

Reply:
`current_motion_x` is being allocated memory and will raise an error. Each function that is called inside the interrupt routine, must accept memory buffers and write the output inside the memory buffer. The memory buffer itself must be pre-allocated in the class `__init__()` method. refer to: https://github.com/BeNeuroLab/PyTreadmillTask/blob/fbd9c407c530d26e7e458601b9f2faa9ed4414b1/devices/_PMW3360DM.py#L267C8-L268C66